pkg/autohealing/cloudprovider/openstack/provider.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// getStackName finds the name of a stack matching a given ID.
+func (provider *OpenStackCloudProvider) getStackName(stackID string) (string, error) {
+	stack, err := stacks.Find(provider.Heat, stackID).Extract()
+	if err != nil {
+		return "", err
+	}
+
+	return stack.Name, nil
+}
+
+// getAllStackResourceMapping returns all mapping relationships including
+// masters and minions(workers). The key in the map is the server/instance ID
+// in Nova and the value is the resource ID and name of the server, and the
+// parent stack ID and name.
+func (provider *OpenStackCloudProvider) getAllStackResourceMapping(stackName, stackID string) (m map[string]ResourceStackRelationship, err error) {
+	if provider.ResourceStackMapping != nil {
+		return provider.ResourceStackMapping, nil
+	}
+
+	mapping := make(map[string]ResourceStackRelationship)
+
+	serverPages, err := stackresources.List(provider.Heat, stackName, stackID, stackresources.ListOpts{Depth: 2}).AllPages()
+	if err != nil {
+		return m, err
+	}
+	serverResources, err := stackresources.ExtractResources(serverPages)
+	if err != nil {
+		return m, err
+	}
+
+	for _, sr := range serverResources {
+		if sr.Type != "OS::Nova::Server" {
+			continue
+		}
+
+		for _, link := range sr.Links {
+			if link.Rel == "self" {
+				continue
+			}
+
+			paths := strings.Split(link.Href, "/")
+			if len(paths) > 2 {
+				m := ResourceStackRelationship{
+					ResourceID:   sr.PhysicalID,
+					ResourceName: sr.Name,
+					StackID:      paths[len(paths)-1],
+					StackName:    paths[len(paths)-2],
+				}
+
+				log.V(4).Infof("resource ID: %s, resource name: %s, parent stack ID: %s, parent stack name: %s", sr.PhysicalID, sr.Name, paths[len(paths)-1], paths[len(paths)-2])
+
+				mapping[sr.PhysicalID] = m
+			}
+		}
+	}
+	provider.ResourceStackMapping = mapping
+
+	return provider.ResourceStackMapping, nil
+}
+
+// Repair soft deletes the VMs, marks the heat resource "unhealthy" then trigger Heat stack update in order to rebuild
+// the VMs. The information this function needs:
+// - Nova VM IDs
+// - Heat stack ID and resource ID.
+func (provider OpenStackCloudProvider) Repair(nodes []healthcheck.NodeInfo) error {
+	clusterName := provider.Config.ClusterName
+	cluster, err := clusters.Get(provider.Magnum, clusterName).Extract()
+	if err != nil {
+		return fmt.Errorf("failed to get the cluster %s, error: %v", clusterName, err)
+	}
+
+	clusterStackName, err := provider.getStackName(cluster.StackID)
+	if err != nil {
+		return fmt.Errorf("failed to get the Heat stack for cluster %s, error: %v", clusterName, err)
+	}
+
+	// In order to rebuild the nodes by Heat stack update, we need to know the parent stack ID of the resources and
+	// mark them unhealthy first.
+	allMapping, err := provider.getAllStackResourceMapping(clusterStackName, cluster.StackID)
+	if err != nil {
+		return fmt.Errorf("failed to get the resource stack mapping for cluster %s, error: %v", clusterName, err)
+	}
+
+	for _, n := range nodes {
```

Comment:
might i suggest batch size and maybe false positives detection, in a network/partial failure scenario, this will be deleting all the nodes.

Reply:
Yes, batch size repair is on our roadmap. I'm waiting for this service to be merged and we could add the support in Magnum, so that we could do the integration test in our preprod.

Reply:
nice, did i miss it? is the roadmap published somewhere?

Reply:
I will submit the feature request after the service is approved, otherwise, I don't think it makes sense to discuss the roadmap for a private service.